packages/lib/isOutOfBounds.tsx

```diff
@@ -17,50 +21,253 @@ function guardAgainstBookingInThePast(date: Date) {
+        countNonBusinessDays: periodCountCalendarDays,
+      });
+
+      return { rollingEndDay, rangeStartDay: null, rangeEndDay: null };
+    }
+
+    case PeriodType.RANGE: {
+      const rangeStartDay = dayjs(periodStartDate).utcOffset(utcOffset).endOf("day");
+      const rangeEndDay = dayjs(periodEndDate).utcOffset(utcOffset).endOf("day");
+      return {
+        rollingEndDay: null,
+        rangeStartDay,
+        rangeEndDay,
+      };
+    }
+  }
+  return {
+    rollingEndDay: null,
+    rangeStartDay: null,
+    rangeEndDay: null,
+  };
+}
+
+/**
+ * TODO: Unit Test this function
+ */
+export function getRollingWindowEndDate({
+  startDate,
+  daysNeeded,
+  allDatesWithBookabilityStatus,
+  countNonBusinessDays,
+}: {
+  /**
+   * It should be already in the correct timezone. No, utcOffset conversion is done in this function
+   */
+  startDate: dayjs.Dayjs;
+  daysNeeded: number;
+  allDatesWithBookabilityStatus: Record<string, { isBookable: boolean }>;
+  countNonBusinessDays: boolean | null;
+}) {
+  const log = logger.getSubLogger({ prefix: ["getRollingWindowEndDate"] });
+  log.debug("called:", safeStringify({ startDay: startDate.format(), daysNeeded }));
+  let counter = 1;
+  let rollingEndDay;
+  let currentDate = startDate.startOf("day");
+
+  // It helps to break out of the loop if we don't find enough bookable days.
+  const maxDaysToCheck = ROLLING_WINDOW_PERIOD_MAX_DAYS_TO_CHECK;
+
+  let bookableDaysCount = 0;
+  // Add periodDays to currentDate, skipping non-bookable days.
+  while (bookableDaysCount < daysNeeded) {
+    // What if we don't find any bookable days. We should break out of the loop after a certain number of days.
+    if (counter > maxDaysToCheck) {
+      break;
+    }
+
+    const isBookable = !!allDatesWithBookabilityStatus[currentDate.format("YYYY-MM-DD")]?.isBookable;
+
+    if (isBookable) {
+      bookableDaysCount++;
+      rollingEndDay = currentDate;
+    }
+
+    log.silly(
+      `Loop Iteration: ${counter}`,
+      safeStringify({
+        currentDate: currentDate.format(),
+        isBookable,
+        bookableDaysCount,
+        rollingEndDay: rollingEndDay?.format(),
+      })
+    );
+
+    currentDate = countNonBusinessDays ? currentDate.add(1, "days") : currentDate.businessDaysAdd(1);
+
+    counter++;
+  }
+
+  /**
+   * We can't just return null(if rollingEndDay couldn't be obtained) and allow days farther than ROLLING_WINDOW_PERIOD_MAX_DAYS_TO_CHECK to be booked as that would mean there is no future limit
+   */
+  const rollingEndDayOrLastPossibleDayAsPerLimit = rollingEndDay ?? currentDate;
+  log.debug("Returning rollingEndDay", rollingEndDayOrLastPossibleDayAsPerLimit.format());
+  return rollingEndDayOrLastPossibleDayAsPerLimit;
+}
+
+/**
+ * To be used when we work on Timeslots(and not Dates) to check boundaries
+ * It ensures that the time isn't in the past and also checks if the time is within the minimum booking notice.
+ */
+export function isTimeOutOfBounds({
```

Comment:
Separated isOutOfBounds in two functions.
`isTimeOutOfBounds` to operate on time slot(to be sed by minimumBookingNotice which works on timeslot) and `isDateOutOfBounds` to operate on a date(to be used by periodType boundary checks)